game/final_obs_gen.py

```diff
@@ -0,0 +1,494 @@
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0],
+                [1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1],
+                [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                [0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                [0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0],
+                [0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0],
+                [0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+            ]
+        )
+
+    def setup_parameters(self):
+
+        # Converts user ratios for the pygame GUI display
+        self.JUMP_HEIGHT = self.U_JUMP_HEIGHT * self.GRID_SIZE
+
+        # Block shape and size
+        self.height, self.width = self.JUMP_ARRAY.shape  # all the same size
+        self.OG_DINO_Y = self.REF_AXIS - self.height * self.GRID_SIZE
+
+        # Dino Jump - Calculates time related values so it makes sense physically
+        self.TIME_TO_PEAK = self.JUMP_TIME / 2
+        self.ACCEL = (2 * self.JUMP_HEIGHT) / (self.TIME_TO_PEAK**2)
+        self.INIT_V = self.TIME_TO_PEAK * self.ACCEL
+
+    ###################################
+
+    def check_jumping(self):
+        # Check if jumping
+
+        # NOTE: JUMP_TIME is given in seconds, whereas pygame records time in milliseconds.
+        # This conversion must be kept in mind throughout the coding process
+        if pygame.time.get_ticks() - self.jump_start_time <= self.JUMP_TIME * 1000:
+            jump_time = (
+                pygame.time.get_ticks() - self.jump_start_time
+            ) / 1000  # There may be a more efficient way to do this
+            # self.dino_y = self.OG_DINO_Y - self.INIT_V * jump_time + self.ACCEL * jump_time**2
+            self.y = self.OG_DINO_Y - self.jumping_coefficient * (
+                jump_time**2 - self.JUMP_TIME * jump_time
+            )  # Tried using the ACCEL and INIT_V function, but it was a little funky

```

Comment:
I believe this calculation is not "pixelated"? ie. `self.y` sometimes is not a multiple of `GRID_SIZE`. This is another reason why it might make sense to avoid calculating the actual y coordinate on the screen, and instead calculate the y coordinate on the "pixel panel".

Reply:
The GRID SIZE thing was confusing, we made a change to this system now!